chainlink-modsec/libmodsec/internal/attestation-indexer/in_memory_indexer.go

```diff
@@ -0,0 +1,125 @@
+func NewInMemoryAttestationIndexer() *InMemoryAttestationIndexer {
+	// Create a new InMemoryAttestationIndexer with an initialized attestations map
+	a := &InMemoryAttestationIndexer{
+		attestations: syncmap.Map{},
+	}
+	return a
+}
+
+// GetAttestationsForMessage retrieves attestations for a given message and converts it into the modsectypes.Attestation type
+func (at *InMemoryAttestationIndexer) GetAttestationsForMessage(_ context.Context, msg modsectypes.Message) ([]modsectypes.Attestation, error) {
+	// Retrieve the map of attestations for the given messageID
+	verifierTypesToIDs, ok := at.attestations.Load(msg.Header.MessageID)
+	if !ok {
+		return nil, nil // No attestations found for this messageID
+	}
+	attestations := make([]modsectypes.Attestation, 0)
+
+	// Iterate over the verifier types for this messageID
+	verifierTypesToIDs.(*syncmap.Map).Range(func(verifierType, verifierIdsMap interface{}) bool {
+		verifierIdsMap.(*syncmap.Map).Range(func(verifierId, proof interface{}) bool {
+			// Append the attestation to the list
+			attestations = append(attestations, modsectypes.Attestation{
+				Proof:      proof.([]byte),
+				VerifierId: verifierId.([32]byte),
+			})
+			return true // Continue iterating over verifierIdsMap
+		})
+		return true
+	})
+
+	return attestations, nil
+}
+
+func (at *InMemoryAttestationIndexer) MessageIsReady(ctx context.Context, msg modsectypes.Message) bool {
+	verifierTypesReady := make([]bool, 0)
+	// we should be looping over all verifier types in the message, but that functionality isn't available yet
+	for _, verifierType := range []modsectypes.VerifierType{modsectypes.VerifierTypeCLCommit} {
+		verifierTypesReady = append(verifierTypesReady, at.VerifierTypeMeetsQuorum(ctx, msg.Header.MessageID, verifierType, 1))
+	}
+
+	// perform an && on all results
+	return !slices.Contains(verifierTypesReady, false)
+}
+
+// AddAttestation adds an attestation for a given messageID and verifierID
+// This is meant to be called as a hook whenever a verifier pushes to their storage
+func (at *InMemoryAttestationIndexer) AddAttestation(_ context.Context, msgId [32]byte, verifierId [32]byte, proof []byte) error {
+	verifierType, err := at.GetVerifierType(verifierId)
+	if err != nil {
+		return fmt.Errorf("failed to get verifier type for verifier %x: %w", verifierId, err)
+	}
+	// If this is the first attestation for this messageID, we need to create a new map for verifier types
+	msgToVerifierTypes, _ := at.attestations.LoadOrStore(msgId, &syncmap.Map{})
+
+	// if this is the first attestation for this verifiertype, we need to create a new map for verifierId
+	verifierTypesToVerifierIds, _ := msgToVerifierTypes.(*syncmap.Map).LoadOrStore(verifierType, &syncmap.Map{})
+
+	verifierTypesToVerifierIds.(*syncmap.Map).Store(verifierId, proof)
+	return nil
+}
+
+func (at *InMemoryAttestationIndexer) VerifierTypeMeetsQuorum(_ context.Context, msgId [32]byte, verifierType modsectypes.VerifierType, required int) bool {
```

Comment:
I think we need to revise `verifierType` - in my mind it should not be used in the API and only verifier ID should be used.